tensornetwork/backends/jax/jitted_functions.py

```diff
@@ -149,3 +150,151 @@ def body_vector(i, vals):
+    return [vector, krylov_vectors, n, H]
+
+  @partial(jax.jit, static_argnums=(5, 6, 7))
+  def _arnoldi_fact(matvec, args, v0, krylov_vectors, H, start, num_krylov_vecs,
+                    eps):
+    """
+    Compute an m-step arnoldi factorization of `matvec`, with 
+    m = min(`it`,`num_krylov_vecs`). The factorization will 
+    do at most `num_krylov_vecs` steps. The returned arrays 
+    `kv` and `H` will satisfy the Arnoldi recurrence relation
+    ```
+    matrix @ Vm - Vm @ Hm - fm * em = 0
+    ```
+    with `matrix` the matrix representation of `matvec` and 
+    `Vm =  jax.numpy.transpose(kv[:it, :])`, 
+    `Hm = H[:it, :it]`, `fm = np.expand_dims(kv[it, :] * H[it, it - 1]`,1) 
+    and `em` a kartesian basis vector of shape `(1, kv.shape[1])` 
+    with `em[0, -1] == 1` and 0 elsewhere.
+
+    Args:
+      matvec: The matrix vector product.
+      args: List of arguments to `matvec`.
+      v0: Initial state to `matvec`.
+      krylov_vectors: An array for storing the krylov vectors. The individual
+        vectors are stored as columns.a
+      H: Matrix of overlaps.
+      start: Integer denoting the start position where the first 
+        produced krylov_vector should be inserted into `krylov_vectors`
+      num_krylov_vecs: Number of krylov iterations, should be identical to 
+        `krylov_vectors.shape[0]`
+      eps: Convergence parameter. Iteration is terminated if the norm of a
+        krylov-vector falls below `eps`.
+    Returns:
+      kv: An array of krylov vectors
+      H: A matrix of overlaps 
+      it: The number of performed iterations.
+    """
+    Z = jax.numpy.linalg.norm(v0)
+    v = v0 / Z
+    krylov_vectors = jax.ops.index_update(krylov_vectors,
+                                          jax.ops.index[start, :],
+                                          jax.numpy.ravel(v))
+    H = jax.lax.cond(
+        start > 0, start,
+        lambda x: jax.ops.index_update(H, jax.ops.index[x, x - 1], Z), None,
+        lambda x: H)
+
+    def body(vals):
+      krylov_vectors, H, matvec, vector, _, threshold, i, maxiter = vals
+      Av = matvec(vector, *args)
+      initial_vals = [Av, krylov_vectors, i, H]
+      Av, krylov_vectors, _, H = jax.lax.fori_loop(
+          0, i + 1, modified_gram_schmidt_step_arnoldi, initial_vals)
+      norm = jax.numpy.linalg.norm(jax.numpy.ravel(Av))
+      Av /= norm
+      H = jax.ops.index_update(H, jax.ops.index[i + 1, i], norm)
+      krylov_vectors = jax.ops.index_update(krylov_vectors,
+                                            jax.ops.index[i + 1, :],
+                                            jax.numpy.ravel(Av))
+      return [krylov_vectors, H, matvec, Av, norm, threshold, i + 1, maxiter]
+
+    def cond_fun(vals):
+      _, _, _, _, norm, threshold, iteration, maxiter = vals
+
+      def check_thresh(check_vals):
+        val, thresh = check_vals
+        return jax.lax.cond(val < thresh, False, lambda x: x, True, lambda x: x)
```

Comment:
These might deserve a block comment spelling out what they do? Not critical though.